File: go/dijkstra/dijkstra_basico.go
package main

import (
	"fmt"
	"math"
)

// SimpleGraph representa um grafo simples com matriz de adjacência
type SimpleGraph struct {
	vertices int
	matrix   [][]int // Matriz de adjacência com pesos (0 = sem conexão)
}

// NewSimpleGraph cria um novo grafo simples
func NewSimpleGraph(vertices int) *SimpleGraph {
	matrix := make([][]int, vertices)
	for i := range matrix {
		matrix[i] = make([]int, vertices)
	}

	return &SimpleGraph{
		vertices: vertices,
		matrix:   matrix,
	}
}

// AddEdge adiciona uma aresta com peso
func (g *SimpleGraph) AddEdge(from, to, weight int) {
	g.matrix[from][to] = weight
}

// DijkstraBasico implementa o algoritmo de Dijkstra de forma didática
// Complexidade: O(V²) onde V é o número de vértices
func (g *SimpleGraph) DijkstraBasico(start int) ([]int, []int) {
	const INF = math.MaxInt32 // Representa infinito

	// Arrays para armazenar resultados
	distances := make([]int, g.vertices) // Menor distância até cada vértice
	previous := make([]int, g.vertices)  // Vértice anterior no caminho mais curto
	visited := make([]bool, g.vertices)  // Controla quais vértices já foram processados

	// Inicialização
	fmt.Println("=== Inicialização ===")
	for i := 0; i < g.vertices; i++ {
		distances[i] = INF // Todas as distâncias começam como infinito
		previous[i] = -1   // Nenhum vértice anterior inicialmente
		visited[i] = false // Nenhum vértice foi visitado
	}
	distances[start] = 0 // Distância do vértice inicial para ele mesmo é 0

	fmt.Printf("Vértice inicial: %d\n", start)
	fmt.Printf("Distâncias iniciais: %v\n\n", distances)

	// Algoritmo principal: processa cada vértice
	for count := 0; count < g.vertices; count++ {
		fmt.Printf("=== Iteração %d ===\n", count+1)

		// Encontra o vértice não visitado com menor distância
		minDistance := INF
		minVertex := -1

		for v := 0; v < g.vertices; v++ {
			if !visited[v] && distances[v] < minDistance {
				minDistance = distances[v]
				minVertex = v
			}
		}

		// Se não encontrou vértice alcançável, termina
		if minVertex == -1 {
			fmt.Println("Não há mais vértices alcançáveis")
			break
		}

		fmt.Printf("Processando vértice %d (distância: %d)\n", minVertex, distances[minVertex])

		// Marca como visitado
		visited[minVertex] = true

		// Relaxa todas as arestas saindo do vértice atual
		fmt.Printf("Verificando vizinhos de %d:\n", minVertex)
		for neighbor := 0; neighbor < g.vertices; neighbor++ {
			// Se existe aresta e o vizinho não foi visitado
			if g.matrix[minVertex][neighbor] > 0 && !visited[neighbor] {
				// Calcula nova distância passando pelo vértice atual
				newDistance := distances[minVertex] + g.matrix[minVertex][neighbor]

				fmt.Printf("  Vizinho %d: distância atual = %d, nova = %d",
					neighbor, distances[neighbor], newDistance)

				// Se a nova distância é melhor, atualiza
				if newDistance < distances[neighbor] {
					distances[neighbor] = newDistance
					previous[neighbor] = minVertex
					fmt.Printf(" -> ATUALIZADO!")
				}
				fmt.Println()
			}
		}

		fmt.Printf("Distâncias após iteração: %v\n\n", distances)
	}

	return distances, previous
}

// GetPath reconstrói o caminho mais curto do start ao end
func (g *SimpleGraph) GetPath(previous []int, start, end int) []int {
	// Se não há caminho para o destino
	if previous[end] == -1 && start != end {
		return nil
	}

	// Reconstrói o caminho de trás para frente
	path := make([]int, 0)
	current := end

	for current != -1 {
		path = append([]int{current}, path...) // Adiciona no início
		current = previous[current]
	}

	return path
}

// DisplayMatrix exibe a matriz de adjacência
func (g *SimpleGraph) DisplayMatrix() {
	fmt.Println("\nMatriz de Adjacência:")
	fmt.Print("     ")
	for i := 0; i < g.vertices; i++ {
		fmt.Printf("%3d ", i)
	}
	fmt.Println()

	for i := 0; i < g.vertices; i++ {
		fmt.Printf("%3d: ", i)
		for j := 0; j < g.vertices; j++ {
			if g.matrix[i][j] == 0 {
				fmt.Print("  - ")
			} else {
				fmt.Printf("%3d ", g.matrix[i][j])
			}
		}
		fmt.Println()
	}
	fmt.Println()
}

// Exemplo de uso
func main() {
	fmt.Println("=== Algoritmo de Dijkstra - Versão Básica ===")

	// Cria um grafo com 5 vértices
	g := NewSimpleGraph(5)

	// Adiciona arestas (origem, destino, peso)
	g.AddEdge(0, 1, 10)
	g.AddEdge(0, 4, 5)
	g.AddEdge(1, 2, 1)
	g.AddEdge(1, 4, 2)
	g.AddEdge(2, 3, 4)
	g.AddEdge(3, 2, 6)
	g.AddEdge(3, 0, 7)
	g.AddEdge(4, 1, 3)
	g.AddEdge(4, 2, 9)
	g.AddEdge(4, 3, 2)

	g.DisplayMatrix()

	// Executa Dijkstra a partir do vértice 0
	distances, previous := g.DijkstraBasico(0)

	// Exibe resultados
	fmt.Println("=== RESULTADOS FINAIS ===")
	fmt.Printf("Vértice de origem: 0\n\n")

	for i := 0; i < g.vertices; i++ {
		if distances[i] == math.MaxInt32 {
			fmt.Printf("Vértice %d: INALCANÇÁVEL\n", i)
		} else {
			path := g.GetPath(previous, 0, i)
			fmt.Printf("Vértice %d: distância = %d, caminho = %v\n",
				i, distances[i], path)
		}
	}

	fmt.Println("\n=== Explicação do Algoritmo ===")
	fmt.Println("1. Inicializa todas as distâncias como infinito, exceto a origem (0)")
	fmt.Println("2. Em cada iteração, escolhe o vértice não visitado com menor distância")
	fmt.Println("3. Para cada vizinho deste vértice, tenta 'relaxar' a aresta:")
	fmt.Println("   - Se (distância atual + peso da aresta) < distância do vizinho:")
	fmt.Println("   - Atualiza a distância do vizinho")
	fmt.Println("4. Marca o vértice como visitado e repete até todos serem processados")
}

File: python/dijkstra/dijkstra_basico.py
"""
Implementação Básica do Algoritmo de Dijkstra
============================================

Esta é uma implementação didática do algoritmo de Dijkstra para encontrar
o caminho mais curto entre dois vértices em um grafo ponderado.

Características desta versão:
- Usa lista simples para a fila de prioridade
- Código mais legível e fácil de entender
- Ideal para aprendizado dos conceitos básicos
- Complexidade: O(V²) onde V é o número de vértices

Autor: Algorithms Repository
Data: Julho 2025
"""

import sys
from typing import Dict, List, Tuple, Optional

class GrafoDijkstra:
    """
    Classe que representa um grafo para o algoritmo de Dijkstra.
    """
    
    def __init__(self):
        """
        Inicializa um grafo vazio.
        """
        self.vertices = {}  # Dicionário de adjacências: {vertice: [(vizinho, peso), ...]}
        
    def adicionar_vertice(self, vertice: str) -> None:
        """
        Adiciona um vértice ao grafo.
        
        Args:
            vertice: Nome do vértice a ser adicionado
        """
        if vertice not in self.vertices:
            self.vertices[vertice] = []
            
    def adicionar_aresta(self, origem: str, destino: str, peso: float) -> None:
        """
        Adiciona uma aresta ponderada ao grafo.
        
        Args:
            origem: Vértice de origem
            destino: Vértice de destino
            peso: Peso da aresta (deve ser não-negativo)
        """
        if peso < 0:
            raise ValueError("O algoritmo de Dijkstra não funciona com pesos negativos")
            
        # Garante que ambos os vértices existem
        self.adicionar_vertice(origem)
        self.adicionar_vertice(destino)
        
        # Adiciona a aresta (grafo direcionado)
        self.vertices[origem].append((destino, peso))
        
    def obter_vertice_menor_distancia(self, distancias: Dict[str, float], visitados: set) -> Optional[str]:
        """
        Encontra o vértice não visitado com menor distância.
        
        Args:
            distancias: Dicionário com as distâncias atuais
            visitados: Conjunto de vértices já visitados
            
        Returns:
            Vértice com menor distância ou None se todos foram visitados
        """
        menor_distancia = sys.maxsize
        vertice_escolhido = None
        
        for vertice in self.vertices:
            if vertice not in visitados and distancias[vertice] < menor_distancia:
                menor_distancia = distancias[vertice]
                vertice_escolhido = vertice
                
        return vertice_escolhido
        
    def dijkstra(self, origem: str, destino: str) -> Tuple[float, List[str]]:
        """
        Executa o algoritmo de Dijkstra para encontrar o caminho mais curto.
        
        Args:
            origem: Vértice de origem
            destino: Vértice de destino
            
        Returns:
            Tupla contendo (distância_total, caminho)
            
        Raises:
            ValueError: Se origem ou destino não existirem no grafo
        """
        if origem not in self.vertices:
            raise ValueError(f"Vértice de origem '{origem}' não existe no grafo")
        if destino not in self.vertices:
            raise ValueError(f"Vértice de destino '{destino}' não existe no grafo")
            
        # Inicialização
        distancias = {vertice: sys.maxsize for vertice in self.vertices}
        distancias[origem] = 0
        
        # Para reconstruir o caminho
        predecessores = {vertice: None for vertice in self.vertices}
        
        # Conjunto de vértices visitados
        visitados = set()
        
        print(f"Iniciando Dijkstra de '{origem}' para '{destino}'")
        print(f"Distâncias iniciais: {distancias}")
        
        # Algoritmo principal
        while len(visitados) < len(self.vertices):
            # Encontra o vértice não visitado com menor distância
            atual = self.obter_vertice_menor_distancia(distancias, visitados)
            
            if atual is None or distancias[atual] == sys.maxsize:
                break  # Não há mais vértices alcançáveis
                
            visitados.add(atual)
            print(f"\nVisitando vértice: {atual} (distância: {distancias[atual]})")
            
            # Se chegamos ao destino, podemos parar
            if atual == destino:
                print("Destino alcançado!")
                break
                
            # Relaxa todas as arestas do vértice atual
            for vizinho, peso in self.vertices[atual]:
                if vizinho not in visitados:
                    nova_distancia = distancias[atual] + peso
                    
                    if nova_distancia < distancias[vizinho]:
                        print(f"  Relaxando {vizinho}: {distancias[vizinho]} -> {nova_distancia}")
                        distancias[vizinho] = nova_distancia
                        predecessores[vizinho] = atual
                        
        # Reconstrói o caminho
        caminho = []
        if distancias[destino] != sys.maxsize:
            atual = destino
            while atual is not None:
                caminho.append(atual)
                atual = predecessores[atual]
            caminho.reverse()
            
        return distancias[destino], caminho
        
    def imprimir_grafo(self) -> None:
        """
        Imprime uma representação do grafo.
        """
        print("Representação do Grafo:")
        for vertice in sorted(self.vertices.keys()):
            vizinhos = ", ".join([f"{v}({p})" for v, p in self.vertices[vertice]])
            print(f"  {vertice} -> {vizinhos if vizinhos else 'sem conexões'}")


def exemplo_basico():
    """
    Exemplo básico de uso do algoritmo de Dijkstra.
    """
    print("=== Exemplo Básico do Algoritmo de Dijkstra ===\n")
    
    # Cria o grafo
    grafo = GrafoDijkstra()
    
    # Adiciona as arestas (representando um mapa de cidades)
    grafo.adicionar_aresta("A", "B", 4)
    grafo.adicionar_aresta("A", "C", 2)
    grafo.adicionar_aresta("B", "C", 1)
    grafo.adicionar_aresta("B", "D", 5)
    grafo.adicionar_aresta("C", "D", 8)
    grafo.adicionar_aresta("C", "E", 10)
    grafo.adicionar_aresta("D", "E", 2)
    
    grafo.imprimir_grafo()
    
    # Executa o algoritmo
    origem = "A"
    destino = "E"
    
    print(f"\nBuscando o caminho mais curto de {origem} para {destino}:")
    print("=" * 50)
    
    distancia, caminho = grafo.dijkstra(origem, destino)
    
    print("\nResultado:")
    if caminho:
        print(f"Caminho: {' -> '.join(caminho)}")
        print(f"Distância total: {distancia}")
    else:
        print("Não há caminho entre os vértices.")


def exemplo_detalhado():
    """
    Exemplo mais detalhado com múltiplas consultas.
    """
    print("\n" + "="*60)
    print("=== Exemplo Detalhado - Rede de Computadores ===")
    print("="*60 + "\n")
    
    # Cria um grafo representando uma rede de computadores
    rede = GrafoDijkstra()
    
    # Adiciona conexões (latência em ms)
    rede.adicionar_aresta("Router1", "PC_A", 1)
    rede.adicionar_aresta("Router1", "Router2", 5)
    rede.adicionar_aresta("Router1", "Router3", 3)
    rede.adicionar_aresta("Router2", "PC_B", 2)
    rede.adicionar_aresta("Router2", "Server", 4)
    rede.adicionar_aresta("Router3", "PC_C", 1)
    rede.adicionar_aresta("Router3", "Server", 6)
    rede.adicionar_aresta("PC_A", "PC_B", 10)
    
    rede.imprimir_grafo()
    
    # Testa múltiplos caminhos
    testes = [
        ("Router1", "Server"),
        ("PC_A", "PC_C"),
        ("PC_B", "PC_C")
    ]
    
    for origem, destino in testes:
        print(f"\nCalculando rota de {origem} para {destino}:")
        print("-" * 40)
        
        distancia, caminho = rede.dijkstra(origem, destino)
        
        if caminho:
            print(f"Rota: {' -> '.join(caminho)}")
            print(f"Latência total: {distancia}ms")
        else:
            print("Sem conectividade entre os dispositivos.")


if __name__ == "__main__":
    exemplo_basico()
    exemplo_detalhado()
